plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/cache.kt

```diff
@@ -39,38 +49,70 @@ class JavaClassCacheManager(val file: File) : Closeable {
+
+                val originSource = javaCache.getSourceForType(aggregatingOrigin)
+                if (originSource.extension == "java") {
+                    sourcesToReprocess.add(originSource)
+                } else if (originSource.extension == "class") {
+                    // This is a generated .class file that we need to reprocess.
+                    classNamesToReprocess.add(aggregatingOrigin)
                 }
+            }
+        }
 
-                SourcesToReprocess.Incremental(
-                    toReprocess.toList(),
-                    generatedDirtyTypes,
-                    aggregatedTypes.also {
-                        it.removeAll(filesToReprocess.dirtyTypes)
-                        it.removeAll(generatedDirtyTypes)
-                        it.removeAll(invalidatedIsolatingId)
-                    })
+        for (impactedType in impactedTypes) {
+            if (impactedType !in isolatingGeneratedTypes && impactedType !in aggregatingGeneratedTypes) {
+                sourcesToReprocess.add(javaCache.getSourceForType(impactedType))
+            } else if (impactedType in isolatingGeneratedTypes) {
+                // this is a generated type by isolating AP
+                val isolatingOrigin = aptCache.getOriginForGeneratedIsolatingType(impactedType, javaCache::getSourceForType)
+                if (isolatingOrigin in impactedTypes || isolatingOrigin in dirtyClasspathFqNames) {
+                    continue
+                }
+                val originSource = javaCache.getSourceForType(isolatingOrigin)
+                if (originSource.extension == "java") {
+                    sourcesToReprocess.add(originSource)
+                } else if (originSource.extension == "class") {
+                    classNamesToReprocess.add(isolatingOrigin)
+                }
             }
         }
+
+        if (sourcesToReprocess.isNotEmpty() || classNamesToReprocess.isNotEmpty()) {
+            // Invalidate state only if there are some files that will be reprocessed
+            javaCache.invalidateDataForTypes(impactedTypes)
+            aptCache.invalidateAggregating()
+            // for isolating, invalidate both own types and classpath types
+            aptCache.invalidateIsolatingForOriginTypes(impactedTypes)
+            aptCache.invalidateIsolatingForOriginTypes(dirtyClasspathFqNames)
+        }
+
+        return SourcesToReprocess.Incremental(sourcesToReprocess.toList(), impactedTypes, classNamesToReprocess)
+    }
+
+    private fun getAllImpactedTypes(changes: Changes, aggregatingGeneratedTypes: Set<String>): MutableSet<String> {
+        val impactedTypes = javaCache.getAllImpactedTypes(changes)
+        // check isolating with origins from the classpath
+        impactedTypes.addAll(aptCache.getIsolatingGeneratedTypesForOrigins(changes.dirtyFqNamesFromClasspath, javaCache::getTypesForFiles))
+        if (changes.sourceChanges.isNotEmpty() || impactedTypes.isNotEmpty()) {
```

Comment:
Could classpath change (e.g. lib dependency added) add additional origins for aggregated AP? How would work kapt in such case?

Reply:
Is classpath change affects only compilation of generated classes or can add additional origins for AP or such origins obtained only within module?

Reply:
Aggregating APs can have origins only within module.

For isolating, it is possible to have origin outside of module as APs are able to specify any originating element. It the origin type from classpath changes, any generated files from it will be removed. However, KAPT will not reprocess that type. It is expected that APs will follow the same type resolution path as they initially did in order to re-generate them. Also, at least one source from module should have type dependency on the classpath type, otherwise it would not be possible for isolating AP to reach it. Supporting these origins is just to make it easier on the isolating APs (see KT-34340).

In general, there are 2 stages in determining impacted types:
a) following type dependencies
b) following annotation processing dependencies

For a), any mention of type `X` in `T` will cause `T` to become invalid if `X` changes. This becomes transitive if `X` is part of `T`s ABI.
For b), if any source/type needs to be reprocessed, we must reprocess all aggregating annotations (as processing any sources/class could generated sources that contain aggregating annotation). Also, all isolating generated types are invalidated if their origins have changed (this is transitive).